### ISSUE TITLE
docker API v2 is not compatible with docker_app_lifecycle.tgz

### ISSUE THREAD

**Issue by ramelito** [cloudfoundry-incubator/docker-registry-release#5]
Environment:
cf v214
diego 0.1412.0
docker-cache 0.1011.0

Reproduce:
1) put image to diego-docker-cache
2) cf docker-push APP diego-docker-cache:8080/IMAGE --no-start
3) cf set-env APP DIEGO_DOCKER_CACHE true
4) cf start APP

builder (from https://github.com/cloudfoundry-incubator/docker_app_lifecycle) tries to pull image from  docker cache via v1 not v2 and here we get stating error

192.168.7.53 - - [03/Sep/2015:07:56:07 +0000] "GET /v1/_ping HTTP/1.1" 404 19 "" "Go 1.1 package http"
time="2015-09-03T07:56:07.862468125Z" level=info msg="response completed" environment=development http.request.host="192.168.8.51:8080" http.request.id=fe734478-66e5-4fbb-a2c2-810ad97890ec http.request.method=GET http.request.remoteaddr="192.168.7.53:37044" http.request.uri="/v1/repositories/altoros/redis/images" http.request.useragent="Go 1.1 package http" http.response.contenttype="text/plain; charset=utf-8" http.response.duration="74.606µs" http.response.status=404 http.response.written=19 instance.id=745d1d5c-6571-420b-a7a4-b388261745d8 service=registry version=v2.0.1 
192.168.7.53 - - [03/Sep/2015:07:56:07 +0000] "GET /v1/repositories/altoros/redis/images HTTP/1.1" 404 19 "" "Go 1.1 package http"

so there should be backward compatibility with builder docker app which is
./docker version
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64

**Comment by hsiliev** [comment]
Our pipeline validates the almost the same start scenario. However we do not push directly to the cache since on start the image would be pulled and then pushed to the cache (thus making the first step redundant).

I wonder if there is some other difference between your environment and our test env. Can you please provide the full log from cf start (`cf logs APP --recent`)?

Also can you elaborate on the first "put image to cache" step? What docker client do you use? How do you tag your image?

**Comment by ramelito** [comment]
Hi, in my scenario i cannot use public hub, only insecure local one. So at the first step i have put image to docker-cache, so builder can get image from insecure docker hub. This is really redundant but this is the only way not to use public hub. Correct me if i am wrong, and there is another option to pull image from insecure docker hub in diego.

> Also can you elaborate on the first "put image to cache" step? What docker client do you use? How do you tag your image?

$ docker --version
Docker version 1.8.1, build d12ea79

**Comment by hsiliev** [comment]
You are right - no way to use insecure registry yet.

Can you create a gist with the logs from the cf start? I'm curious why we have this working in the CI and it fails on your side.

**Comment by ramelito** [comment]
Here is the logs, i am not sure they are verbose enough

```
cf logs redis --recent
Connected, dumping recent logs for app redis in org diego / space diego as admin...

2015-09-03T14:16:48.68+0000 [API/0]      OUT Created app with guid 27cf5023-0b0f-470a-8a68-2f4874b5f451
2015-09-03T14:16:49.08+0000 [API/0]      OUT Updated app with guid 27cf5023-0b0f-470a-8a68-2f4874b5f451 ({"route"=>"1b6fe578-9ed0-48f1-b173-26be8ab17605"})
2015-09-03T14:16:57.27+0000 [API/0]      OUT Updated app with guid 27cf5023-0b0f-470a-8a68-2f4874b5f451 ({"environment_json"=>"PRIVATE DATA HIDDEN"})
2015-09-03T14:17:05.91+0000 [API/0]      OUT Updated app with guid 27cf5023-0b0f-470a-8a68-2f4874b5f451 ({"state"=>"STARTED"})
2015-09-03T14:17:07.19+0000 [API/0]      ERR Failed to stage application: staging failed
```

Is there insecure docker option in the roadmap and if yes, what is the timeline for it?

**Comment by hsiliev** [comment]
Not sure the issue is directly connected with the registry/cache here. The normal logs in case staging failed should read:

```
2015-09-02T15:08:19.25+0000 [API/0]      OUT Created app with guid 941283df-ba05-49e0-be1d-2e79618e3722
2015-09-02T15:08:21.45+0000 [API/0]      OUT Updated app with guid 941283df-ba05-49e0-be1d-2e79618e3722 ({"route"=>"d80362a5-cea6-46a7-bdee-7c7f33b5db7b"})
2015-09-02T15:08:22.32+0000 [API/0]      OUT Updated app with guid 941283df-ba05-49e0-be1d-2e79618e3722 ({"environment_json"=>"PRIVATE DATA HIDDEN"})
2015-09-02T15:08:23.45+0000 [API/0]      OUT Updated app with guid 941283df-ba05-49e0-be1d-2e79618e3722 ({"state"=>"STARTED"})
2015-09-02T15:08:23.53+0000 [STG/0]      OUT Creating container
2015-09-02T15:08:23.96+0000 [STG/0]      OUT Successfully created container
2015-09-02T15:08:24.08+0000 [STG/0]      OUT Staging...
2015-09-02T15:08:24.10+0000 [STG/0]      OUT Staging process started ...
2015-09-02T15:08:24.16+0000 [STG/0]      ERR Staging process failed: Exit trace for group:
... <real error> ...
```

In your case we doesn't seem to have any staging container/task created at all, so I think the root cause is not really Docker client incompatibility. Can you please take a look at Diego's bridge logs? TPS or Stager logs are the most probable places that might shed more light on the problem.

The support for insecure registry should come with https://www.pivotaltracker.com/story/show/100495522. The plan is to have it by the end of the year, but priorities might change.

**Comment by ramelito** [comment]
> In your case we doesn't seem to have any staging container/task created at all, so I think the root cause is not really Docker client incompatibility.

I am not sure logging and reporting working well. Let me show what i am seeing in garden-linux out.log
container created:

```
":{"handle":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f","rootfs":"/var/vcap/packages/rootfs_cflinuxfs2/rootfs","properties":{"executor:allocated-at":"1441289825622034014","executor:cpu-weight":"100","executor:disk-mb":"5120","executor:log-config":"{\"guid\":\"27cf5023-0b0f-470a-8a68-2f4874b5f451\",\"index\":0,\"source_name\":\"STG\"}","executor:memory-mb":"1024","executor:metrics-config":"{\"guid\":\"\",\"index\":0}","executor:owner":"executor","executor:result":"{\"failed\":false,\"failure_reason\":\"\",\"stopped\":false}","executor:rootfs":"/var/vcap/packages/rootfs_cflinuxfs2/rootfs","executor:start-timeout":"0","executor:state":"created","tag:domain":"cf-app-staging","tag:lifecycle":"task","tag:result-file":"/tmp/docker-result/result.json"},"privileged":true,"limits":{"bandwidth_limits":{},"cpu_limits":{},"disk_limits":{},"memory_limits":{}}},"session":"4.3953"}}
```

and spawned with builder

```
{"timestamp":"1441289826.192689896","source":"garden-linux","message":"garden-linux.garden-server.run.spawned","log_level":1,"data":{"handle":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f","id":1,"session":"4.3987","spec":{"path":"/tmp/docker_app_lifecycle/builder","args":["-outputMetadataJSONFilename","/tmp/docker-result/result.json","-dockerRef","192.168.8.51:8080/altoros/redis:latest","-cacheDockerImage","-dockerRegistryAddresses","192.168.8.51:8080","-insecureDockerRegistries","192.168.8.51:8080"],"env":["VCAP_APPLICATION={\"limits\":{\"mem\":1024,\"disk\":1024,\"fds\":16384},\"application_id\":\"27cf5023-0b0f-470a-8a68-2f4874b5f451\",\"application_version\":\"25beb036-376e-4262-a731-62f4fb77e710\",\"application_name\":\"redis\",\"version\":\"25beb036-376e-4262-a731-62f4fb77e710\",\"name\":\"redis\",\"space_name\":\"diego\",\"space_id\":\"5b7d8f0c-b6a7-4d8a-9e91-d8a7845a936a\"}","VCAP_SERVICES={}","MEMORY_LIMIT=1024m","CF_STACK=cflinuxfs2","DIEGO_DOCKER_CACHE=true","CF_INSTANCE_IP=192.168.7.53","CF_INSTANCE_PORT=","CF_INSTANCE_ADDR=","CF_INSTANCE_PORTS="],"user":"root","rlimits":{"nofile":16384}}}}
```

and here logs from Stager

```
{"timestamp":"1441289912.067365885","source":"stager","message":"stager.docker.build-recipe.staging-request","log_level":1,"data":{"app-id":"27cf5023-0b0f-470a-8a68-2f4874b5f451","session":"2.4","staging-guid":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f"}}
{"timestamp":"1441289912.068814278","source":"stager","message":"stager.staging-handler.staging-request.desiring-task","log_level":1,"data":{"callback_url":"http://stager.service.cf.internal:8888/v1/staging/27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f/completed","session":"3.3","staging-guid":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f","task_guid":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f"}}
{"timestamp":"1441289913.327290773","source":"stager","message":"stager.completion-handler.task-complete-callback-received.posting-staging-complete","log_level":1,"data":{"guid":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f","payload":"eyJleGVjdXRpb25fbWV0YWRhdGEiOiIiLCJkZXRlY3RlZF9zdGFydF9jb21tYW5kIjpudWxsLCJlcnJvciI6eyJpZCI6IlN0YWdpbmdFcnJvciIsIm1lc3NhZ2UiOiJzdGFnaW5nIGZhaWxlZCJ9fQ==","session":"4.3"}}{"timestamp":"1441289913.327432871","source":"stager","message":"stager.completion-handler.task-complete-callback-received.cc-client.delivering-staging-response","log_level":1,"data":{"guid":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f","payload":"{\"execution_metadata\":\"\",\"detected_start_command\":null,\"error\":{\"id\":\"StagingError\",\"message\":\"staging failed\"}}","session":"4.3.1"}}
{"timestamp":"1441289913.363405466","source":"stager","message":"stager.completion-handler.task-complete-callback-received.cc-client.delivered-staging-response","log_level":1,"data":{"guid":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f","session":"4.3.1"}}
{"timestamp":"1441289913.363565207","source":"stager","message":"stager.completion-handler.task-complete-callback-received.posted-staging-complete","log_level":1,"data":{"guid":"27cf5023-0b0f-470a-8a68-2f4874b5f451-474956c4ff344db1a2c9ad2ceb8cc16f","session":"4.3"}}
{"timestamp":"1441297395.619236946","source":"stager","message":"stager.docker.build-recipe.staging-request","log_level":1,"data":{"app-id":"3686a39f-7e1a-4e1b-9c5d-056a28307077","session":"2.6","staging-guid":"3686a39f-7e1a-4e1b-9c5d-056a28307077-4fcb73df395d47e6921fb3f9d265575c"}}
{"timestamp":"1441297395.619437218","source":"stager","message":"stager.staging-handler.staging-request.desiring-task","log_level":1,"data":{"callback_url":"http://stager.service.cf.internal:8888/v1/staging/3686a39f-7e1a-4e1b-9c5d-056a28307077-4fcb73df395d47e6921fb3f9d265575c/completed","session":"3.4","staging-guid":"3686a39f-7e1a-4e1b-9c5d-056a28307077-4fcb73df395d47e6921fb3f9d265575c","task_guid":"3686a39f-7e1a-4e1b-9c5d-056a28307077-4fcb73df395d47e6921fb3f9d265575c"}}

```

and no useful information from TPS.

Ok i get information about insecure docker roadmap, so one can definitely exploit docker cache as origin insecure server if during builder stage there will be compatibility between docker API versions. Actually i dont know may it is worth to bump version of docker in docker_app_lifecycle rather than downgrade version of registry in docker cache, or may be have an option to choose between versions of docker cache registry?

**Comment by hsiliev** [comment]
I think I found the root cause of the problem - we don't use the IP of the cache registry anymore, but instead rely on consul.

The only workaround I can think of, is to pushing your image (step 1) with tag including `docker-registry.service.cf.internal:8080` and not IP. You can add hosts entry to trick Docker or do the push from inside CF.

As a solution you can a private registry with trusted/signed SSL cert or wait for the insecure registry support.

**Comment by ramelito** [comment]
> I think I found the root cause of the problem - we don't use the IP of the cache registry anymore, but instead rely on consul.

As far as i can understand, consul is needed to retrieve ip address on a stager phase, right?

> The only workaround I can think of, is to pushing your image (step 1) with tag including docker-registry.service.cf.internal:8080

In fact it is with this tag, but this tags have local significance i suppose.

> As a solution you can a private registry with trusted/signed SSL cert or wait for the insecure registry support.

And could i somehow to hack cflinuxfs2 to inject self-signed certificate?

**Comment by hsiliev** [comment]
Haven't tried this, but it should be possible to add a self-signed cert to the trusted certs in cflinuxfs2. Since docker is written in go you should check what would be the correct way to have this supported by golang.

**Comment by emalm** [comment]
Hi, @ramelito,

We don't intend you as an end user to be pushing images to the diego-docker-cache registry: it's supposed to be only for the internal use of the Diego deployment. That said, it is a perfectly good v2 registry when deployed on its own, and we're fixing support for staging images from v2 registries in https://www.pivotaltracker.com/story/show/105158076. We also now support passing a whitelist of insecure registries to the stager, and we'll make sure that that functionality is preserved as we support the v2 registry API.

Thanks,
Eric

**Comment by emalm** [comment]
The changes to support staging images from v2 registries are included in Diego v0.1439.0, which is recommended for use with CF v223.
